src/metabase/db/custom_migrations.clj

```diff
@@ -1255,13 +1255,12 @@
     (qs/shutdown scheduler)))
 
 (define-migration BackfillQueryField
-  (let [update-query-fields! (requiring-resolve 'metabase.native-query-analyzer/update-query-fields-for-card!)
+  (let [update-query-fields! (requiring-resolve 'metabase.models.query-field/update-query-fields-for-card!)
```

Comment:
Huh. Mmmm. I just noticed this migration uses `update-query-fields-for-card!`, which is outside the custom-migrations namespace and has a high chance of being changed. Migrations can't change and testing them is difficult. If we change the behaviour of a migration we have a serious issue (usually P1), so we have to try to keep their dependencies to a minimum (e.g. avoid using toucan model code, the QP, etc).

See the namespace docstring for more:
https://github.com/metabase/metabase/blob/5b9522cfc9f2dd2c0a136a983fbe2175a67c41f7/src/metabase/db/custom_migrations.clj#L2-L9

Or 

In this case, using the `update-query-fields-for-card!` function couples its code (and its dependencies, including MLv2 and the `query-field` model code) to the specific form that `dataset_query` takes when this migration runs. Because of this migration, it will be very difficult to change `dataset_query` in the future if we wanted to, except in ways that are forward compatible with this migration. And changing `update-query-fields-for-card!` or MLv2 now comes with greater risk of high-impact bugs. Instead of a runtime error when a certain query is being built/executed, we'll have a migration error, probably P1.

https://github.com/metabase/metabase/blob/97211bb4159d7ff6e1accc11c9cbe79f0dbe7fdf/src/metabase/db/custom_migrations.clj#L9

Unfortunately I don't have any insights to how to do this without depending on MLv2 without a ton of work and testing. I'd suggest getting QPD advice, because it will require reimplementing a stripped-down version of the MLv2 functions that you're using and keeping them unchanged in this namespace. The good news is, we won't have to maintain it because it should never change. But we need to know it works, either through test coverage or simplicity.

Reply:
Also, I think this function will only work on normalized MBQL queries (those returned by `maybe-normalized-query`), but will likely not be able to handle all the previous versions of MBQL queries we've stored in the past. I could be wrong but I doubt MLv2 functions normalize all the different forms of clauses we've stored in the past. That logic should be limited to `metabase.legacy-mbql.normalize`, and used from the toucan model code.

Reply:
Updated in last commit.